frontend/src/framework/ModuleInstanceStatusController.ts

```diff
@@ -0,0 +1,87 @@
+import { cloneDeep } from "lodash";
+
+export enum StatusMessageType {
+    Warning = "warning",
+    Error = "error",
+}
+
+export enum StatusSource {
+    View = "view",
+    Settings = "settings",
+}
+
+export type StatusMessage = {
+    source: StatusSource;
+    message: string;
+    type: StatusMessageType;
+};
+
+export type StatusControllerState = {
+    messages: StatusMessage[];
+    loading: boolean;
+    viewDebugMessage: string;
+    settingsDebugMessage: string;
+    viewRenderCount: number | null;
+    settingsRenderCount: number | null;
+};
+
+export class ModuleInstanceStatusController {
```

Comment:
After reading through the code and considering the inheritance and division of labour between `ModuleInstanceStatusController` and `ModuleInstanceStatusControllerPrivate` I think we should approach it in a slightly different way.

I suggest we make `ModuleInstanceStatusController` a pure interface and then defer the entire implementation to a `ModuleInstanceStatusControllerImpl` class. Will be cleaner wrt division of labour and we can lose some of the internal type definitions in `ModuleInstanceStatusController.ts`

Thinking something like this:
```
export interface ModuleInstanceStatusController {
    addMessage(source: StatusSource, message: string, type: StatusMessageType): void;
    clearMessages(source: StatusSource): void;
    setLoading(isLoading: boolean): void;

    setDebugMessage(source: StatusSource, message: string): void;
    incrementReportedComponentRenderCount(source: StatusSource): void;

    reviseAndPublishState(): void;
}
```